scipy/linalg/decomp_cholesky.py

```diff
@@ -2,35 +2,45 @@
 
 from __future__ import division, print_function, absolute_import
 
-from numpy import asarray_chkfinite, asarray
+from numpy import asarray_chkfinite, asarray, atleast_2d
 
 # Local imports
 from .misc import LinAlgError, _datacopied
 from .lapack import get_lapack_funcs
 
 __all__ = ['cholesky', 'cho_factor', 'cho_solve', 'cholesky_banded',
-            'cho_solve_banded']
+           'cho_solve_banded']
 
 
 def _cholesky(a, lower=False, overwrite_a=False, clean=True,
-                check_finite=True):
+              check_finite=True):
     """Common code for cholesky() and cho_factor()."""
 
-    if check_finite:
-        a1 = asarray_chkfinite(a)
-    else:
-        a1 = asarray(a)
-    if len(a1.shape) != 2 or a1.shape[0] != a1.shape[1]:
-        raise ValueError('expected square matrix')
+    a1 = asarray_chkfinite(a) if check_finite else asarray(a)
+    a1 = atleast_2d(a1)
```

Comment:
The upcasting is not consistent. Some upcast (linalg.solve), some don't(cho_solve). Independent from this PR the `linalg` module really needs some love in terms of naming consistency, argument consistency and so on. If you can let me know about your opinion about this upcasting I'll try to follow that as a guideline. I will open a separate issue for such questions but for this specific case, my rationale is as follows:

The user can give array'ble input and we pass it through `asarray()` to convert to have a `ndarray`. At this point, the inputs 
  - 1.0
  - [1.0]
fail. 

Note that, these are only invalid at the scipy level. But direct ?POTRF calls accepts without any problem. Example: 

```
sp.linalg.lapack.zpotrf(2j)
(array(0j), 1)

sp.linalg.lapack.dpotrf([2])
(array([ 1.41421356]), 0)

sp.linalg.lapack.dpotrf(2)
(array(1.4142135623730951), 0)
```
I would argue that this is a regression at the higher level. 

Another use-case I frequently encounter; This also prevents to set up any loops since every loop has to take care of scalar arguments, should any other returns a 1D scalar array. Hence for every loop the user has to implement traps or `try..except..` to handle scalar 1D instances. 

Hence, I think it makes sense to upcast and investigate the provided object for consistency and shape checks. What do you (and whoever else wants to chime in) think?